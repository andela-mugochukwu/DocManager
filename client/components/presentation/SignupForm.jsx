import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'sweetalert2';
import $ from 'jquery';

/**
 * Fires the function that authenticates a user
 * @param {object} event - object containing data about the control that
 * triggered an event
 * @param {func} signUserUp - function that executes when signing users up
 * @param {object} history - the browser and app state history
 * @param {func} getUserDocuments - Function that executes when
   * fetching documents belonging to a particular user
 * @return {null} returns void
 */
const signUp = (event, signUserUp, history, getUserDocuments) => {
  event.preventDefault();
  const userName = $('#signupform input[name=username]').val();
  const email = $('#signupform input[type=email]').val();
  const password = $('#signupform input[type=password]').val();
  const comfirmPassword = $('#signupform input[type=password]').val();
  const roleId = $('#signupform input[name=group1]:checked').val();
  const formData = { userName, email, password, comfirmPassword, roleId };
  signUserUp(formData)
    .then((res) => {
      if (res.status === 'successful') {
        history.push('/user/documents');
        getUserDocuments(res.data.userId, res.data.token);
      } else {
        Alert({
          title: 'Error Signing up',
          text: res.message,
          type: 'error',
          confirmButtonText: 'ok'
        });
      }
    });
};
/**
 * Renders form for signin up new user
 * @return {object} Returns the signup form to render
 */
const SignupForm = ({ signUserUp, history,
  signUpButtonText, getUserDocuments }) => (
    <div id="signupform" style={{ display: 'none' }}>
      <input
        className="forminput"
        name="username"
        type="text"
        minLength="3"
        placeholder="Username"
        required
      />
      <input
        className="forminput"
        name="email"
        type="email"
        minLength="10"
        placeholder="Email"
        required
      />
      <input
        className="forminput"
        name="password"
        type="password"
        minLength="6"
        placeholder="Password"
        required
      />
      <input
        className="forminput"
        name="comfirmpassword"
        type="password"
        placeholder="Comfirm password"
        required
      />
      <div id="roles">
        Role:&nbsp;&nbsp;
        <input
          className="with-gap"
          name="group1"
          type="radio"
          id="Learning"
          value="2"
        />
        <label htmlFor="Learning">Learning</label>
        <input
          className="with-gap"
          name="group1"
          type="radio"
          id="Fellow"
          value="1"
        />
        <label htmlFor="Fellow">Fellow</label>
        <input
          className="with-gap"
          name="group1"
          type="radio"
          id="DevOps"
          value="3"
        />
        <label htmlFor="DevOps">DevOps</label><br />
      </div>
      <button
        id="signinbtn"
        className="center-align waves-effect waves-light btn"
        onClick={event => signUp(event, signUserUp, history, getUserDocuments)}
      >{signUpButtonText}</button>
    </div>
);

SignupForm.propTypes = {
  signUserUp: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  signUpButtonText: PropTypes.string.isRequired,
  getUserDocuments: PropTypes.func.isRequired,
};
export default SignupForm;
