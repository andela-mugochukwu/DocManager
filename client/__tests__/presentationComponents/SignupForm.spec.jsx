import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import SignupForm from '../../components/presentation/SignupForm';

describe('The signupform component:', () => {
  const history = { push: url => url };
  const props = {
    signUserUp: jest.fn(() => Promise.resolve()),
    allDocuments: jest.fn(() => Promise.resolve()),
    submitButton: 'submit',
    roleType: 'Admin',
    history,
  };
  const enzymeWrapper = mount(<SignupForm {...props} />);
  test('should render the signup form correctly', () => {
    const signUserUp = jest.fn(() => Promise.resolve());
    const allDocuments = jest.fn(() => Promise.resolve());
    const submitButton = 'submit';
    const roleType = 'Admin';
    const component = renderer.create(
      <SignupForm
        history={history}
        signUserUp={signUserUp}
        submitButton={submitButton}
        roleType={roleType}
        allDocuments={allDocuments}
      />
  );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should change submit button when it is clicked', () => {
    const event = {
      preventDefault: () => 'hello',
    };
    enzymeWrapper.find('button#signupbtn').props().onClick(event);
    expect(enzymeWrapper.find('button#signupbtn').text()).toBe('submit');
  });
});
