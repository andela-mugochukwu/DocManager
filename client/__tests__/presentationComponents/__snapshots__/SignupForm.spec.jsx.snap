// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The signupform component: should render the signup form correctly 1`] = `
<div
  id="signupform"
  style={
    Object {
      "display": "none",
    }
  }
>
  <input
    className="forminput"
    minLength="3"
    name="username"
    placeholder="Username"
    required={true}
    type="text"
  />
  <input
    className="forminput"
    minLength="10"
    name="email"
    placeholder="Email"
    required={true}
    type="email"
  />
  <input
    className="forminput"
    minLength="6"
    name="password"
    placeholder="Password"
    required={true}
    type="password"
  />
  <input
    className="forminput"
    name="comfirmpassword"
    placeholder="Comfirm password"
    required={true}
    type="password"
  />
  <div
    id="roles"
  >
    Role:  
    <input
      className="with-gap"
      data-value="Learning"
      id="Learning"
      name="group1"
      type="radio"
      value="3"
    />
    <label
      htmlFor="Learning"
    >
      Learning
    </label>
    <input
      className="with-gap"
      data-value="Fellow"
      id="Fellow"
      name="group1"
      type="radio"
      value="2"
    />
    <label
      htmlFor="Fellow"
    >
      Fellow
    </label>
    <input
      className="with-gap"
      data-value="DevOps"
      id="DevOps"
      name="group1"
      type="radio"
      value="4"
    />
    <label
      htmlFor="DevOps"
    >
      DevOps
    </label>
    <br />
  </div>
  <button
    className="center-align waves-effect waves-light btn"
    id="signupbtn"
    onClick={[Function]}
  >
    submit
  </button>
</div>
`;
